dotcom-rendering/src/amp/components/ContentABTest.tsx

```diff
@@ -0,0 +1,94 @@
+import React from 'react';
+import sha256 from 'crypto-js/sha256';
+
+type ContentABTestGroup = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11;
+
+interface ContentABTestContext {
+	group?: ContentABTestGroup;
+}
+
+const NUM_GROUPS = 12;
+
+const isContentABTestGroup = (n: number): n is ContentABTestGroup =>
+	n === 0 ||
+	n === 1 ||
+	n === 2 ||
+	n === 3 ||
+	n === 4 ||
+	n === 5 ||
+	n === 6 ||
+	n === 7 ||
+	n === 8 ||
+	n === 9 ||
+	n === 10 ||
+	n === 11;
+
```

Comment:
At the risk of bikeshedding could we simplify the above to:

```typescript
const AB_TEST_GROUPS = [0,1,2,3,4,5,6,7,8,9,10,11] as const;
type ContentABTestGroup = typeof AB_TEST_GROUPS[number];
const NUM_GROUPS = AB_TEST_GROUPS.length;

const isContentABTestGroup = (n: ContentABTestGroup): n is ContentABTestGroup =>  AB_TEST_GROUPS.includes(n);
```

Reply:
And looking at this isContentABTestGroup looks like it's doing nothing. But the generated js does do the check

Reply:
Think that makes a lot of sense.

Only thing is we'd probably want

```tsx
const isContentABTestGroup = (n: number): n is ContentABTestGroup => { };
```
since it needs to be able to narrow any `number` to `ContentABTestGroup`. So we'd maybe end up with something like

```tsx
const isContentABTestGroup = (n: number): n is ContentABTestGroup =>
    AB_TEST_GROUPS.includes(n as ContentABTestGroup);
```

do you think that's preferable?

Reply:
Fine by me. I _think_ it shakes out the same in the end.

At runtime it's not really narrowing a _number_ to a group id. It's checking if the variable is a valid groupid. Both in my version and this one.

In your version when you're writing it typescript can tell you that only numbers are acceptable to this function. In my version when you're writing it typescript can tell you that only 0-11 are acceptable

In both cases we'll have to tell typescript that the output of hash % 12 is an integer 0-11, and it's a valid ContentABTestGroup (It'll just know it's a number).

Reply:
They come out (identical)[https://www.typescriptlang.org/play?#code/MYewdgzgLgBAggIQPoBUCiBlFSDiAlAeQFUAFDGAXhgG0AGAGgEZ6AmegZnoBZ6BWegGz0A7PQAc9AJxMGjRgF0YAQwgxQkKAG4AUFACeABwCmMAMLgoRsFEQoj0HACcQAVwOUY+4yABm8ZOhYuISkGNRgLgC2AEZGjvI66tAwAHJEALLBxGQeiKiY2PjZGAB0ADZWAOZQABY62kmwAJYQ5tZWNgh2Ds5uHgAUYABcZhYdtvZQTq4GAJQjYDAto+3WEz0zlAB8MP75QUWhJU1gwGUuACb2g7OJ4MktbZZrXZPTbiwDwzARMXHzPyWqie41eGz6FC22l2uzygUKITIx1O5yuEEGymBYxe3SmvTm9SAA]